File: src/bun_alloc/BufferFallbackAllocator.zig
/// An allocator that attempts to allocate from a provided buffer first,
/// falling back to another allocator when the buffer is exhausted.
/// Unlike `std.heap.StackFallbackAllocator`, this does not own the buffer.
const BufferFallbackAllocator = @This();

#fallback_allocator: Allocator,
#fixed_buffer_allocator: FixedBufferAllocator,

pub fn init(buffer: []u8, fallback_allocator: Allocator) BufferFallbackAllocator {
    return .{
        .#fallback_allocator = fallback_allocator,
        .#fixed_buffer_allocator = FixedBufferAllocator.init(buffer),
    };
}

pub fn allocator(self: *BufferFallbackAllocator) Allocator {
    return .{
        .ptr = self,
        .vtable = &.{
            .alloc = alloc,
            .resize = resize,
            .remap = remap,
            .free = free,
        },
    };
}

fn alloc(ctx: *anyopaque, len: usize, alignment: std.mem.Alignment, ra: usize) ?[*]u8 {
    const self: *BufferFallbackAllocator = @ptrCast(@alignCast(ctx));
    return FixedBufferAllocator.alloc(
        &self.#fixed_buffer_allocator,
        len,
        alignment,
        ra,
    ) orelse self.#fallback_allocator.rawAlloc(len, alignment, ra);
}

fn resize(ctx: *anyopaque, buf: []u8, alignment: std.mem.Alignment, new_len: usize, ra: usize) bool {
    const self: *BufferFallbackAllocator = @ptrCast(@alignCast(ctx));
    if (self.#fixed_buffer_allocator.ownsPtr(buf.ptr)) {
        return FixedBufferAllocator.resize(
            &self.#fixed_buffer_allocator,
            buf,
            alignment,
            new_len,
            ra,
        );
    }
    return self.#fallback_allocator.rawResize(buf, alignment, new_len, ra);
}

fn remap(ctx: *anyopaque, memory: []u8, alignment: std.mem.Alignment, new_len: usize, ra: usize) ?[*]u8 {
    const self: *BufferFallbackAllocator = @ptrCast(@alignCast(ctx));
    if (self.#fixed_buffer_allocator.ownsPtr(memory.ptr)) {
        return FixedBufferAllocator.remap(
            &self.#fixed_buffer_allocator,
            memory,
            alignment,
            new_len,
            ra,
        );
    }
    return self.#fallback_allocator.rawRemap(memory, alignment, new_len, ra);
}

fn free(ctx: *anyopaque, buf: []u8, alignment: std.mem.Alignment, ra: usize) void {
    const self: *BufferFallbackAllocator = @ptrCast(@alignCast(ctx));
    if (self.#fixed_buffer_allocator.ownsPtr(buf.ptr)) {
        return FixedBufferAllocator.free(
            &self.#fixed_buffer_allocator,
            buf,
            alignment,
            ra,
        );
    }
    return self.#fallback_allocator.rawFree(buf, alignment, ra);
}

pub fn reset(self: *BufferFallbackAllocator) void {
    self.#fixed_buffer_allocator.reset();
}

const std = @import("std");
const Allocator = std.mem.Allocator;
const FixedBufferAllocator = std.heap.FixedBufferAllocator;

File: src/bun_alloc/BufferFallbackAllocator.rs
//! An allocator that attempts to allocate from a provided buffer first,
//! falling back to another allocator when the buffer is exhausted.
//! The buffer is borrowed, not owned.

use core::ffi::c_void;

use crate::{Alignment, AllocatorVTable, FixedBufferAllocator, StdAllocator};

pub struct BufferFallbackAllocator<'a> {
    fallback: StdAllocator,
    fixed: FixedBufferAllocator<'a>,
}

impl<'a> BufferFallbackAllocator<'a> {
    pub fn init(buffer: &'a mut [u8], fallback: StdAllocator) -> BufferFallbackAllocator<'a> {
        BufferFallbackAllocator {
            fallback,
            fixed: FixedBufferAllocator::init(buffer),
        }
    }

    pub fn allocator(&self) -> StdAllocator {
        StdAllocator {
            ptr: std::ptr::from_ref::<Self>(self).cast_mut().cast::<c_void>(),
            vtable: &VTABLE,
        }
    }

    pub fn reset(&self) {
        self.fixed.reset();
    }
}

static VTABLE: AllocatorVTable = AllocatorVTable {
    alloc,
    resize,
    remap,
    free,
};

unsafe fn alloc(ctx: *mut c_void, len: usize, alignment: Alignment, ra: usize) -> *mut u8 {
    // SAFETY: ctx was set to `&BufferFallbackAllocator` in `allocator()`.
    let self_: &BufferFallbackAllocator = unsafe { &*ctx.cast::<BufferFallbackAllocator>() };
    self_
        .fixed
        .alloc(len, alignment, ra)
        .or_else(|| self_.fallback.raw_alloc(len, alignment, ra))
        .unwrap_or(core::ptr::null_mut())
}

unsafe fn resize(
    ctx: *mut c_void,
    buf: &mut [u8],
    alignment: Alignment,
    new_len: usize,
    ra: usize,
) -> bool {
    // SAFETY: ctx was set to `&BufferFallbackAllocator` in `allocator()`.
    let self_: &BufferFallbackAllocator = unsafe { &*ctx.cast::<BufferFallbackAllocator>() };
    if self_.fixed.owns_ptr(buf.as_ptr()) {
        return self_.fixed.resize(buf, alignment, new_len, ra);
    }
    self_.fallback.raw_resize(buf, alignment, new_len, ra)
}

unsafe fn remap(
    ctx: *mut c_void,
    memory: &mut [u8],
    alignment: Alignment,
    new_len: usize,
    ra: usize,
) -> *mut u8 {
    // SAFETY: ctx was set to `&BufferFallbackAllocator` in `allocator()`.
    let self_: &BufferFallbackAllocator = unsafe { &*ctx.cast::<BufferFallbackAllocator>() };
    if self_.fixed.owns_ptr(memory.as_ptr()) {
        return self_
            .fixed
            .remap(memory, alignment, new_len, ra)
            .unwrap_or(core::ptr::null_mut());
    }
    self_
        .fallback
        .raw_remap(memory, alignment, new_len, ra)
        .unwrap_or(core::ptr::null_mut())
}

unsafe fn free(ctx: *mut c_void, buf: &mut [u8], alignment: Alignment, ra: usize) {
    // SAFETY: ctx was set to `&BufferFallbackAllocator` in `allocator()`.
    let self_: &BufferFallbackAllocator = unsafe { &*ctx.cast::<BufferFallbackAllocator>() };
    if self_.fixed.owns_ptr(buf.as_ptr()) {
        return self_.fixed.free(buf, alignment, ra);
    }
    self_.fallback.raw_free(buf, alignment, ra)
}
